core/lib/node/src/node/mod.rs

```diff
@@ -0,0 +1,269 @@
+use std::{any::Any, cell::RefCell, collections::HashMap, fmt};
+
+use futures::{future::BoxFuture, FutureExt};
+use tokio::{runtime::Runtime, sync::watch};
+use zksync_health_check::CheckHealth;
+
+pub use self::{context::NodeContext, stop_receiver::StopReceiver};
+use crate::{
+    resource::ResourceProvider,
+    task::{TaskInitError, ZkSyncTask},
+};
+
+mod context;
+mod stop_receiver;
+
+type TaskConstructor =
+    Box<dyn FnOnce(&NodeContext<'_>) -> Result<Box<dyn ZkSyncTask>, TaskInitError>>;
+type HealthCheckTaskConstructor = Box<
+    dyn FnOnce(
+        &NodeContext<'_>,
+        Vec<Box<dyn CheckHealth>>,
+    ) -> Result<Box<dyn ZkSyncTask>, TaskInitError>,
+>;
+
+/// "Manager" class of the node. Collects all the resources and tasks,
+/// then runs tasks until completion.
+///
+/// Initialization flow:
+/// - Node instance is created with access to the resource provider.
+/// - Task constructors are added to the node. At this step, tasks are not created yet.
+/// - Optionally, a healthcheck task constructor is also added.
+/// - Once the `run` method is invoked, node
+///   - attempts to create every task. If there are no tasks, or at least task
+///     constructor fails, the node will return an error.
+///   - initializes the healthcheck task if it's provided.
+///   - waits for any of the tasks to finish.
+///   - sends stop signal to all the tasks.
+///   - waits for the remaining tasks to finish.
+///   - calls `after_node_shutdown` hook for every task that has provided it.
+///   - returns the result of the task that has finished.
+pub struct ZkSyncNode {
+    /// Primary source of resources for tasks.
+    resource_provider: Box<dyn ResourceProvider>,
+    /// Cache of resources that have been requested at least by one task.
+    resources: RefCell<HashMap<String, Box<dyn Any>>>,
+    /// List of task constructors.
+    task_constructors: Vec<(String, TaskConstructor)>,
+    /// Optional constructor for the healthcheck task.
+    healthcheck_task_constructor: Option<HealthCheckTaskConstructor>,
+
+    /// Sender used to stop the tasks.
+    stop_sender: watch::Sender<bool>,
+    /// Stop receiver to be shared with tasks.
```

Comment:
you can always subscribe to watch::Sender, there is no need to keep the receiver around.